spec/components/crop-content/CropContent.spec.tsx

```diff
@@ -0,0 +1,84 @@
+      inventore veritatis et quasi architecto beatae vitae dicta sunt explicabo.
+      Nemo enim ipsam voluptatem quia voluptas sit aspernatur aut odit aut
+      fugit, sed quia consequuntur magni dolores eos qui ratione voluptatem
+      sequi nesciunt. Neque porro quisquam est, qui dolorem ipsum quia dolor sit
+      amet, consectetur, adipisci velit, sed quia non numquam eius modi tempora
+      incidunt ut labore et dolore magnam aliquam quaerat voluptatem. Ut enim ad
+      minima veniam, quis nostrum exercitationem ullam corporis suscipit
+      laboriosam, nisi ut aliquid ex ea commodi consequatur? Quis autem vel eum
+      iure reprehenderit qui in ea voluptate velit esse quam nihil molestiae
+      consequatur, vel illum qui dolorem eum fugiat quo voluptas nulla pariatur?
+    </CropContent>
+  );
+
+const getToggleLink = wrapper => {
+  return wrapper.find('.tk-crop-content .toggle-container a');
+};
+
+const getToggleContainer = wrapper => {
+  return wrapper.find('.tk-crop-content .toggle-container');
+};
+const getContentBlock = wrapper => {
+  return wrapper.find('.tk-crop-content .content');
+};
+describe('CropContent Component', () => {
+  describe('should be created properly', () => {
+    it('render with default props does not crash', () => {
+      const wrapper = shallow(<CropContent></CropContent>);
+      expect(wrapper.length).toEqual(1);
+      expect(wrapper.hasClass('tk-crop-content')).toBe(true);
+    });
+    it('content should crop at 80px by default', () => {
+      const wrapper = getWrapper();
+      wrapper.setState({ hasOverflow: true });
+      expect(getContentBlock(wrapper).prop('style').maxHeight).toBe('80px');
+      expect(getToggleContainer(wrapper).length).toBe(1);
+    });
+    it('toggle should appear according to overflow state', () => {
+      const wrapper = getWrapper();
+      expect(getToggleContainer(wrapper).length).toBe(0);
+      wrapper.setState({ hasOverflow: true });
+      expect(getToggleContainer(wrapper).length).toBe(1);
+    });
+    it('toggle should switch state and height', () => {
+      const wrapper = getWrapper();
+      wrapper.setState({ hasOverflow: true });
+      expect(getToggleContainer(wrapper).length).toBe(1);
+      getToggleLink(wrapper).simulate('click');
+      expect(getContentBlock(wrapper).prop('style').maxHeight).toBe('100vh');
+    });
+    it('should handle overflow on cropHeight change', () => {
+      const wrapper = getWrapper();
+      const handleOverflowSpy = jest.spyOn(
+        wrapper.instance() as CropContent,
+        'handleOverflow'
+      );
+      wrapper.setProps({ cropHeight: 100 });
+      expect(handleOverflowSpy).toHaveBeenCalled();
+    });
+    it('should handle overflow only on collapsed state', () => {
+      const wrapper = getWrapper();
+      wrapper.setState({ hasOverflow: true });
+      getToggleLink(wrapper).simulate('click');
+
+      const setStateSpy = jest.spyOn(wrapper.instance(), 'setState');
```

Comment:
I'm not a fan of spying on internal methods that much, not on calling methods on the instance directly. If your implementation changes, you'll have to change the tests.
Couldn't you for ex assert that `state.hasOverflow` hasn't changed, rather than checking that setState is not getting called ?

Reply:
Agree

Reply:
Hi @micaste ,
Actually the test is focusing more on: did we tried to update the hasOverflow? 
Our component's behaviour will do only if it's in collapsed state otherwise it doesn't.
Knowing if the value's changing is kind of impossible since it depends on the containers rendered size (which is not an easy task to do with enzyme, correct me if I'm wrong)

Reply:
> did we tried to update the hasOverflow?

Why does this matter ? I would only be concerned by the question _"has the `hasOverflow` value changed ?"_
I would agree with your question if you told me _"that's for optimization, to verify if we don't call `setState` too many times"_ to which I would answer  _"react optimizes for you, so you don't really care, and also you're testing too much internals IMO"_
So if you want to answer the question _"has the `hasOverflow` value changed ?"_, you can do `expect(getToggleLink(wrapper)).toBe(undefined)` => actually better than testing the internal state.

> Knowing if the value's changing is kind of impossible since it depends on the containers rendered size

You can do
```
wrapper.instance().containerElRef = {
   offsetHeight: ...,
   scrollHeight: ...,
}
wrapper.find(ResizeDetectDiv).prop("onWidthChange")()
expect(getToggleLink(wrapper)).toBe(undefined)
```

For example.
I'm not hard set on this, but just trying to give you more ideas on what's possible. I usually try to test components "from the outside", not really calling the methods and the states. IMO it makes the tests more robust and needing less updating.
Also, you might want to look into `mount`, that might make the resizing work out of the box, and you might just have to click and assert on node sizes or presence in the dom.

For me the point of unit-tests is to be able to test edge-cases that are otherwise hard to test when using the component in its "functional usage". But if from the outside, without mocking anything, you manage to test all the branches, then I don't see a need to test the internals of calls to the state and things like that.


For example, I see that your method `getWrapper` does not accept props. This means that throughout your unit-test you are not testing your component using different props. Maybe if you did, you wouldn't have to use `wrapper.setState()` etc

Reply:
I don't like testing with internal calls either, thanks for the explanation I'm pretty new to enzyme , need to dig more ^^

Reply:
I used your suggestion to fake size instead of forcing the result through state ;)